Sources/Composable.swift

```diff
@@ -275,7 +275,8 @@ extension NSMutableAttributedString {
             return
         }
 
-        let lastCharacterRange = NSRange(location: length - 1, length: 1)
+        let lastCharacterLength = string.suffix(1).utf16.count
```

Comment:
Pretty sure `suffix` is a O(n) operation. I'm wondering if there's something we can do something like `.reversed.first` that will count from the end and still get the extended grapheme clusters right?

Reply:
@ZevEisenberg I just run some tests and it looks that `.suffix`, `.reversed.first` and `last` are all O(1) operations. I couldn't find any official documentation about it, though.

On the other hand, now that I though about it, I would probably go for something like
```guard length != 0, let lastCharacter = string.last else {
            return
        }

        let lastCharacterLength = String(lastCharacter).utf16.count
        let lastCharacterRange = NSRange(location: length - lastCharacterLength, length: lastCharacterLength)```

just because it looks less hackish than using suffix. But maybe I'm overthinking it.

Reply:
@ZevEisenberg I just run some tests and it looks that .suffix, .reversed.first and last are all O(1) operations. I couldn't find any official documentation about it, though.

The syntax advantage of suffix is that it is automatically coalesced with an empty string, while for first and last we would have to unwrap it (even though we already know the string is longer than 0 characters)

Reply:
You are correct. Not sure what I was thinking. `suffix` iterates backwards through the UTF-8 or UTF-16 code points (depending on Swift version) until it finds the start of a `Character`, so it will be O(number of unicode scalars of last `Character`). Or something. I’m probably misusing the term for the things it’s looping through, but that’s the gist.